pkg/archive/changes.go

```diff
@@ -123,8 +127,16 @@ func Changes(layers []string, rw string) ([]Change, error) {
 							return nil
 						}
 					}
+					changedDirs[path] = struct{}{}
 					change.Kind = ChangeModify
 					break
+				} else if os.IsNotExist(err) && f.Mode().IsRegular() {
```

Comment:
Hi @duglin, yes we're trying to fix the case where `f.Size() == stat.Size() && f.Mode() == stat.Mode() && sameFsTime(f.ModTime(), stat.ModTime())` is true but there are indeed new files created in the RW layer. This might happen when the RW layer is created immediately after (or rather nearly the same time) that the RO layer is created (see the patch that I tried to apply to changes_test.go for details).

Before opening this PR I did the following experiments on ext4 filesystem.
```
root@local-dev:~# mkdir A
root@local-dev:~# stat -c '%y' A
2015-06-03 10:53:39.969511999 +0800
root@local-dev:~# mkdir A/B
root@local-dev:~# stat -c '%y' A
2015-06-03 10:53:48.009511999 +0800
root@local-dev:~# stat -c '%y' A/B
2015-06-03 10:53:48.009511999 +0800
root@local-dev:~# echo test >A/B/file.txt
root@local-dev:~# stat -c '%y' A/B
2015-06-03 10:54:07.193511999 +0800  # modify time changed
root@local-dev:~# stat -c '%y' A
2015-06-03 10:53:48.009511999 +0800  # unchanged
```
From the test results above we can see that when the new file A/B/file.txt is created, only the direct parent directory B is modified. The modify time of directory A is unchanged. So I think it's not necessary to recursively add all the parent directories to `changedDirs`.